cli/mender-artifact/copy.go

```diff
@@ -117,6 +118,48 @@ func doCopy(c *cli.Context, repack *bool, r *io.ReadCloser,
 	return nil
 }
 
+// Install installs a file from the host filesystem onto either
+// a mender artifact, or an sdimg.
+func Install(c *cli.Context) (err error) {
+	fmt.Println("Installing...")
+	var r io.ReadCloser
+	var w PartitionReadWriteClosePacker
+	switch parseCLIOptions(c) {
+	case copyin:
+		var perm os.FileMode
+		if c.String("mode") == "" {
+			return cli.NewExitError("File permissions needs to be set, if you are simply copying, the cp command should fit your needs", 1)
+		}
+		var i int
+		i, err = strconv.Atoi(c.String("mode"))
+		if err != nil {
+			return cli.NewExitError(fmt.Sprintf("%v", err), 1)
+		}
+		perm = os.FileMode(i)
+		r, err = os.OpenFile(c.Args().First(), os.O_RDWR, perm)
+		if err != nil {
+			return cli.NewExitError(fmt.Sprintf("%v", err), 1)
+		}
+		w, err = NewPartitionFile(c.Args().Get(1), c.String("key"))
+		if err != nil {
+			return cli.NewExitError(fmt.Sprintf("%v", err), 1)
+		}
+		if _, err = io.Copy(w, r); err != nil {
+			return cli.NewExitError(fmt.Sprintf("%v", err), 1)
+		}
+		if err = w.Repack(); err != nil {
```

Comment:
I would think that between `Repack` and the previous `if`, you should have:
1. An explicit `w.Close()`. Buffering can cause `w` to not be fully up to date from `debugfs`'s perspective unless you either flush or close the filehandle.
2. `Fchmod` on the `w` handle, which is where you set the permissions. Am I correct that `debugfs` grabs the permissions from the file you try to put in the image?

Reply:
1) debugfs has long since finished it's magic, when close is called on w. The close simply closes the tmp-files that are buffering the debugfs-data.
2) That is correct!

Reply:
Hmm, ok, I think it's obvious that I'm not following the logic here. I'll trust you then! :-)

Reply:
I don't like that. Let's have a quick 2 min on it tomorrow.